### PR TITLE
🐛Fix form-validators with date picker on touch devices

### LINKED ISSUE #22688: amp-date-picker keyboard popup, amp-form fails to validate
URL: https://www.cheapoair.com/flights/amp

When a user inputs a depart or return date into the `amp-date-picker`, the keyboard pops up and covers the calendar:

GjC44dBGpmZ

(Note: a keyboard will not popup in Chrome emulation, need to try on physical device)

The issue happens on this page (page #1): https://www.cheapoair.com/flights/amp
But **does not** happen on this page (page #2): https://www.cheapoair.com/flights/cheap-flights-from-new-york-city-to-bangkok-nyc-bkk/amp

The difference is that there is a `readonly` attribute in the input field on the page where the issue is happening (page #1):

```
<input placeholder="Date"
       id="fromDt"
       autocomplete="off"
       class="date"
       name="fromDt"
       [disabled]="formDataVar.oneway"
       required=""
       readonly=""   <----------- READONLY ATTRIBUTE
/>    
```

Removing the `readonly` attribute fixes the keyboard popup issue, but allows the user to submit the `amp-form` without depart and return dates, which are required input fields. The user should instead see a validation message that looks like this: 

DED86OpfWNk

### PR DESCRIPTION
**Pull request by cvializ**
Fixes #22688

Instead of assigning `readonly` at load-time, assign it just before the mobile device focuses on the input element. This prevents the keyboard from appearing and allows validation to continue as normal.

### CODE REVIEW

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 1]
```diff
@@ -982,12 +984,49 @@ export class AmpDatePicker extends AMP.BaseElement {
...
 
+  /**
+   * Handle mousedown to prevent the mobile keyboard
+   * @param {!Event} e
+   * @private
+   */
+  addTouchReadonly_(e) {
+    const target = dev().assertElement(e.target);
+
+    if (this.isDateField_(target) && isNonValidationReadonly(target)) {
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: Use early return instead

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 2]
```diff
@@ -1922,6 +1961,33 @@ export class AmpDatePicker extends AMP.BaseElement {
...
+  if (!toggle) {
+    delete element.dataset[AMP_READONLY_DATA_ATTR];
+    return;
+  }
+
+  element.dataset[AMP_READONLY_DATA_ATTR] = true;
+}
+
+/**
+ *
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: empty line

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 3]
```diff
@@ -1922,6 +1961,33 @@ export class AmpDatePicker extends AMP.BaseElement {
   }
 }
 
+const AMP_READONLY_DATA_ATTR = 'iAmphtmlReadonly';
+
+const AMP_DATE_BLUR_DATA_ATTR = 'iAmphtmlDateBlur';
+
+/**
+ *
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: empty line

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 4]
```diff
@@ -1922,6 +1961,33 @@ export class AmpDatePicker extends AMP.BaseElement {
...
+ * @param {!Element} element
+ * @param {boolean} toggle
+ */
+function setNonValidationReadonly(element, toggle) {
+  if (!toggle) {
+    delete element.dataset[AMP_READONLY_DATA_ATTR];
+    return;
+  }
+
+  element.dataset[AMP_READONLY_DATA_ATTR] = true;
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: Would add a comment that says the actual `readonly` is set when the element gets focus.

File: extensions/amp-date-picker/0.1/amp-date-picker.js [hunk 5]
```diff
@@ -982,12 +984,49 @@ export class AmpDatePicker extends AMP.BaseElement {
     }
   }
 
+  /**
+   * Handle mousedown to prevent the mobile keyboard
+   * @param {!Event} e
+   * @private
+   */
+  addTouchReadonly_(e) {
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/amp-date-picker.js]: The naming is confusing here, then function name says touch, the comment says mouse but it seems to apply to both (as well as other sources of focus like keyboard).

Comment (cvializ) [reply extensions/amp-date-picker/0.1/amp-date-picker.js]: Changed the naming elsewhere to clarify

File: extensions/amp-date-picker/0.1/amp-date-picker.js (hunk 4 above)
Comment (cvializ) [reply extensions/amp-date-picker/0.1/amp-date-picker.js]: Added comment to the method body

File: extensions/amp-date-picker/0.1/amp-date-picker.js (hunk 1 above)
Comment (cvializ) [reply extensions/amp-date-picker/0.1/amp-date-picker.js]: Nice, that cleaned it up better

File: extensions/amp-date-picker/0.1/test/test-amp-date-picker.js [hunk 6]
```diff
@@ -433,21 +433,75 @@ describes.realWin(
...
+            sandbox.stub(picker.input_, 'isTouchDetected').returns(true);
+
+            return picker
+              .buildCallback()
+              .then(() => {
+                expect(input).to.have.attribute('data-i-amphtml-readonly');
+                return picker.layoutCallback();
+              })
+              .then(() => {
+                const fakeEvent = {target: input};
```
Comment (sparhami) [inline extensions/amp-date-picker/0.1/test/test-amp-date-picker.js]: Would it be possible to instead do this via a call to `.focus()`? That would also test the protection against running the blur handler when you blur/refocus.

Comment (cvializ) [reply extensions/amp-date-picker/0.1/test/test-amp-date-picker.js]: I initially tried that strategy, but it felt like something better reserved for an e2e test. I'll make an issue to test that in a followup #22844

**Comment by mdiblasio** [comment]
hello, when can we expect this fix to go live? the issue is rather urgent as it's negatively impacting the user experience on our AMP pages.

**Comment by cvializ** [comment]
It should be live as of [release 1906191709480](https://github.com/ampproject/amphtml/releases/tag/1906191709480)

**Comment by Sitanderkumar** [comment]
thanks, so what is the solution approach what changes we need to do in our existing code.

**Comment by cvializ** [comment]
Existing AMP markup shouldn't need to change. This PR's js code adds the `readonly` at a different time to stop the keyboard from appearing.

**Comment by Sitanderkumar** [comment]
keyboard overlay still coming after click on date-picker field.

https://m.cheapoair.com/flights/amp

you can find same in above URL, check this url on real device chrome emulator not giving such issue.

**Comment by cvializ** [comment]
`touch-keyboard-editable` is still present on the amp-date-picker tag in the markup delivered by the server.

```

<!-- ROUND-TRIP DATE PICKER I -->
<div class="container">
<amp-date-picker type="range" id="daterange" on="select:AMP.setState({dates: event.dates,startDate: event.start,endDate: event.end})" mode="overlay" locale="en" format="MM-DD-Y" number-of-months="1" start-input-selector="[name=fromDt]" end-input-selector="#toDt" minimum-nights="0" touch-keyboard-editable="">
  <div class="ampstart-input inline-block mt1">
	<span>Depart</span>
	<input placeholder="Date" id="fromDt" autocomplete="off" class="date" name="fromDt" [disabled]="formDataVar.oneway" required="" />
  </div>
  <div class="ampstart-input inline-block mt1 right " id="returnDate">
	<span>Return</span>
	<input placeholder="Date" id="toDt" autocomplete="off" class="date right" name="toDt" [disabled]="formDataVar.oneway" required="" />
  </div>
  <!-- <template type="amp-mustache" date-template="" default="">
	{{^isBlocked}}{{^isHighlighted}}{{D}}{{/isHighlighted}}{{/isBlocked}}{{#isBlocked}}:({{/isBlocked}}{{#isHighlighted}}{{D}}!{{/isHighlighted}}
  </template> -->
  <template type="amp-mustache" info-template="">
    <span [text]="(startDate && endDate ? 'You picked ' + startDate.date + ' as start date and ' + endDate.date + ' as end date.' : 'You will see your chosen dates here.')">
      You will see your chosen dates here.
    </span>
  </template>
</amp-date-picker>

<!-- ONE-WAY DATE PICKER -->
<amp-date-picker type="single" hidden="" id="singleDatePicker" on=" select: AMP.setState({dates: event.dates, tacoTuesdays: event.dates.filter(d=> d.id=='tacos2').length,})" mode="overlay" locale="en" format="MM-DD-Y" number-of-months="1" input-selector="#fromDt-oneway" touch-keyboard-editable="">
  <div class="ampstart-input inline-block mt1">
	<span>Depart</span>
	<input placeholder="Date" name="fromDt" autocomplete="off" id="fromDt-oneway" class="date" disabled="" [disabled]="!formDataVar.oneway" required="" />
  </div>
  <!-- <template type="amp-mustache" date-template="" default="">
	{{^isBlocked}}{{^isHighlighted}}{{D}}{{/isHighlighted}}{{/isBlocked}}{{#isBlocked}}:({{/isBlocked}}{{#isHighlighted}}{{D}}!{{/isHighlighted}}
  </template> -->
  <template type="amp-mustache" info-template="">
    <span [text]="(startDate && endDate ? 'You picked ' + startDate.date + ' as start date and ' + endDate.date + ' as end date.' : 'You will see your chosen dates here.')">
      You will see your chosen dates here.
    </span>
  </template>
</amp-date-picker>
```

That attribute needs to be removed to stop the keyboard from displaying

**Comment by cvializ** [comment]
I'm going to lock this PR for commenting. If you're still experiencing the issue, feel free to comment on the original issue #22688

**Pull request merged**